### PR TITLE
Clipboard keys

### PR DESCRIPTION
**Pull request by willmcgugan**
- Implements cut/copy/paste with ctrl+x, ctrl+c, and ctrl+v
- Binds quit to ctrl+q (previously ctrl+c)

### CODE REVIEW

File: src/textual/widgets/_input.py [hunk 1]
```diff
@@ -995,6 +997,21 @@ async def action_submit(self) -> None:
         )
         self.post_message(self.Submitted(self, self.value, validation_result))
 
-    def action_copy_selection(self) -> None:
+    def action_cut(self) -> None:
+        """Cut the current selection (copy to clipboard and remove from input)."""
+        start, end = sorted(self.selection)
+        text = self.value[start:end]
+        self.app.copy_to_clipboard(text)
```
Comment (darrenburns) [inline src/textual/widgets/_input.py:1004]: I think there's a property "selected_text" for this.

File: src/textual/widgets/_input.py [hunk 2]
```diff
@@ -995,6 +997,21 @@ async def action_submit(self) -> None:
...
         """Copy the current selection to the clipboard."""
         self.app.copy_to_clipboard(self.selected_text)
+
+    def action_paste(self) -> None:
+        """Paste from the local clipboard."""
+        clipboard = self.app._clipboard
+        start, end = sorted(self.selection)
+        new_value = self.value[:start] + clipboard + self.value[end:]
+        self.value = new_value
+        self.cursor_position = start + len(clipboard)
```
Comment (darrenburns) [inline src/textual/widgets/_input.py:1017]: Does "replace" or "insert_at_cursor" work for this as a one liner?

File: src/textual/widgets/_text_area.py [hunk 3]
```diff
@@ -2199,6 +2202,33 @@ def action_delete_line(self) -> None:
...
             self.move_cursor_relative(columns=end_column, record_width=False)
 
+    def action_cut(self) -> None:
+        """Cut text (remove and copy to clipboard)."""
+        if self.read_only:
+            return
+        start, end = self.selection
+        if start == end:
+            return
+        copy_text = self.get_text_range(start, end)
```
Comment (darrenburns) [inline src/textual/widgets/_text_area.py:2212]: I think the "delete_via_keyboard" method returns the text that was deleted as part of the ExitResult, so you shouldn't need to get the range again here.

See "_on_paste"

**Review by darrenburns — APPROVED** [review]
Just minor things

File: tests/text_area/test_textarea_cut_copy_paste.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+async def test_copy():
+    """Check that copy places text in the clipboard."""
+    app = TextAreaApp()
+    async with app.run_test() as pilot:
+        text_area = app.query_one(TextArea)
+        await pilot.click(text_area)
+        await pilot.press(*"Hello, World")
+        await pilot.press("left", "shift+left", "shift+left")
+        await pilot.press("ctrl+c")
+        assert text_area.document.text == "Hello, World"
```
Comment (darrenburns) [inline tests/text_area/test_textarea_cut_copy_paste.py:32]: NP: I'm pretty sure you can do "text_area.text"

File: src/textual/widgets/_input.py [hunk 5]
```diff
@@ -995,6 +997,17 @@ async def action_submit(self) -> None:
...
+        self.delete_selection()
+
+    def action_copy(self) -> None:
         """Copy the current selection to the clipboard."""
         self.app.copy_to_clipboard(self.selected_text)
+
+    def action_paste(self) -> None:
+        """Paste from the local clipboard."""
+        clipboard = self.app._clipboard
+        start, end = sorted(self.selection)
```
Comment (darrenburns) [inline src/textual/widgets/_input.py:1012]: Also no need to sort here I believe.

**Comment by darrenburns** [comment]
I tested and "command + x" does cut the line for me in VSCode and on an online version of VSCode. I think we should revert `ctrl+x` with no selection to cut the line instead of it doing nothing.

**Comment by willmcgugan** [comment]
It doesn't seem to cut the line for me. Does nothing AFAICT.

Not apposed to changing it.

**Comment by darrenburns** [comment]
You must have some config/plugin/keybind override going on as I tested it on an online VSCode with no config and it cut. I'll PR that as it'll just take a moment.

**Comment by darrenburns** [comment]
Turns out the new binding wasn't working anyway:

```python
Binding("ctrl+shift+x", "delete_line", "Delete line", show=False),
```

Should be `ctrl+X` :)

**Comment by willmcgugan** [comment]
Really? `textual keys` reports `ctrl+shift+x`

**Comment by darrenburns** [comment]
Uh oh
image

**Comment by willmcgugan** [comment]
Very strange! is that running in VSCode?

**Pull request merged**